File: backend/services/module-service.go
package services

import (
	"context"
	"fmt"
	"time"

	"backend/models"
	"backend/repository"

	"go.mongodb.org/mongo-driver/bson/primitive"
)

type ModuleService interface {
	GetAllModules(ctx context.Context, req *models.GetModulesRequest) (*models.GetModulesResponse, error)
	GetModuleByID(ctx context.Context, moduleID primitive.ObjectID) (*models.Module, error)
	CreateModule(ctx context.Context, req *models.CreateModuleRequest, userID primitive.ObjectID) (*models.Module, error)
	UpdateModule(ctx context.Context, moduleID primitive.ObjectID, req *models.UpdateModuleRequest, userID primitive.ObjectID) (*models.Module, error)
	DeleteModule(ctx context.Context, moduleID primitive.ObjectID) error
	ToggleModulePublication(ctx context.Context, moduleID primitive.ObjectID, published bool, userID primitive.ObjectID) (*models.Module, error)
	ReorderModules(ctx context.Context, moduleIDs []string, userID primitive.ObjectID) error
	ReorderSubModules(ctx context.Context, moduleID primitive.ObjectID, subModuleIDs []string, userID primitive.ObjectID) error
	BulkReorder(ctx context.Context, req *models.BulkReorderRequest, userID primitive.ObjectID) error

	// SubModule methods
	CreateSubModule(ctx context.Context, moduleID primitive.ObjectID, req *models.CreateSubModuleRequest, userID primitive.ObjectID) (*models.SubModule, error)
	UpdateSubModule(ctx context.Context, moduleID primitive.ObjectID, subModuleID primitive.ObjectID, req *models.CreateSubModuleRequest, userID primitive.ObjectID) (*models.SubModule, error)
	DeleteSubModule(ctx context.Context, moduleID primitive.ObjectID, subModuleID primitive.ObjectID) error
	ToggleSubModulePublication(ctx context.Context, moduleID primitive.ObjectID, subModuleID primitive.ObjectID, published bool, userID primitive.ObjectID) (*models.SubModule, error)
}

type moduleService struct {
	moduleRepo repository.ModuleRepository
}

func NewModuleService(moduleRepo repository.ModuleRepository) ModuleService {
	return &moduleService{
		moduleRepo: moduleRepo,
	}
}

func (s *moduleService) GetAllModules(ctx context.Context, req *models.GetModulesRequest) (*models.GetModulesResponse, error) {
	// Set defaults
	if req.Page < 1 {
		req.Page = 1
	}
	if req.Limit < 1 {
		req.Limit = 10
	}
	if req.Limit > 100 {
		req.Limit = 100
	}

	modules, total, err := s.moduleRepo.GetAllModules(ctx, req)
	if err != nil {
		return nil, fmt.Errorf("failed to get modules: %w", err)
	}

	totalPages := int(total) / req.Limit
	if int(total)%req.Limit > 0 {
		totalPages++
	}

	return &models.GetModulesResponse{
		Modules:    modules,
		Total:      total,
		Page:       req.Page,
		Limit:      req.Limit,
		TotalPages: totalPages,
	}, nil
}

func (s *moduleService) GetModuleByID(ctx context.Context, moduleID primitive.ObjectID) (*models.Module, error) {
	module, err := s.moduleRepo.GetModuleByID(ctx, moduleID)
	if err != nil {
		return nil, fmt.Errorf("module not found: %w", err)
	}

	return module, nil
}

func (s *moduleService) CreateModule(ctx context.Context, req *models.CreateModuleRequest, userID primitive.ObjectID) (*models.Module, error) {
	now := time.Now()

	// If no order specified, set it to a high value (will be auto-ordered by creation time)
	order := req.Order
	if order == 0 {
		// Get the count of existing modules to set default order
		// This ensures new modules without explicit order go to the end
		allModules, _, err := s.moduleRepo.GetAllModules(ctx, &models.GetModulesRequest{Page: 1, Limit: 1000})
		if err == nil {
			order = len(allModules) + 1
		} else {
			order = 999 // Fallback if query fails
		}
	}

	module := &models.Module{
		ID:          primitive.NewObjectID(),
		Name:        req.Name,
		Description: req.Description,
		Content:     req.Content,
		SubModules:  []models.SubModule{},
		IsPublished: false, // Always start as draft
		Order:       order,
		CreatedAt:   now,
		UpdatedAt:   now,
		CreatedBy:   userID,
		UpdatedBy:   userID,
	}

	// Create submodules if provided
	if len(req.SubModules) > 0 {
		for i, subModuleReq := range req.SubModules {
			subModuleOrder := subModuleReq.Order
			if subModuleOrder == 0 {
				subModuleOrder = i + 1 // Auto-order submodules 1, 2, 3...
			}

			subModule := models.SubModule{
				ID:          primitive.NewObjectID(),
				Name:        subModuleReq.Name,
				Description: subModuleReq.Description,
				Content:     subModuleReq.Content,
				IsPublished: false, // Always start as draft
				Order:       subModuleOrder,
				CreatedAt:   now,
				UpdatedAt:   now,
				CreatedBy:   userID,
				UpdatedBy:   userID,
			}
			module.SubModules = append(module.SubModules, subModule)
		}
	}

	if err := s.moduleRepo.CreateModule(ctx, module); err != nil {
		return nil, fmt.Errorf("failed to create module: %w", err)
	}

	return module, nil
}

func (s *moduleService) UpdateModule(ctx context.Context, moduleID primitive.ObjectID, req *models.UpdateModuleRequest, userID primitive.ObjectID) (*models.Module, error) {
	// Get existing module
	module, err := s.moduleRepo.GetModuleByID(ctx, moduleID)
	if err != nil {
		return nil, fmt.Errorf("module not found: %w", err)
	}

	// Update fields if provided
	if req.Name != nil {
		module.Name = *req.Name
	}
	if req.Description != nil {
		module.Description = *req.Description
	}
	if req.Content != nil {
		module.Content = *req.Content
	}
	if req.Order != nil {
		module.Order = *req.Order
	}
	if req.SubModules != nil {
		module.SubModules = req.SubModules
	}

	module.UpdatedAt = time.Now()
	module.UpdatedBy = userID

	if err := s.moduleRepo.UpdateModule(ctx, module); err != nil {
		return nil, fmt.Errorf("failed to update module: %w", err)
	}

	return module, nil
}

func (s *moduleService) DeleteModule(ctx context.Context, moduleID primitive.ObjectID) error {
	if err := s.moduleRepo.DeleteModule(ctx, moduleID); err != nil {
		return fmt.Errorf("failed to delete module: %w", err)
	}
	return nil
}

func (s *moduleService) ToggleModulePublication(ctx context.Context, moduleID primitive.ObjectID, published bool, userID primitive.ObjectID) (*models.Module, error) {
	module, err := s.moduleRepo.GetModuleByID(ctx, moduleID)
	if err != nil {
		return nil, fmt.Errorf("module not found: %w", err)
	}

	module.IsPublished = published
	module.UpdatedAt = time.Now()
	module.UpdatedBy = userID

	if err := s.moduleRepo.UpdateModule(ctx, module); err != nil {
		return nil, fmt.Errorf("failed to update module publication status: %w", err)
	}

	return module, nil
}

// SubModule methods

func (s *moduleService) CreateSubModule(ctx context.Context, moduleID primitive.ObjectID, req *models.CreateSubModuleRequest, userID primitive.ObjectID) (*models.SubModule, error) {
	module, err := s.moduleRepo.GetModuleByID(ctx, moduleID)
	if err != nil {
		return nil, fmt.Errorf("module not found: %w", err)
	}

	now := time.Now()

	// If no order specified, set it to be after existing submodules
	order := req.Order
	if order == 0 {
		order = len(module.SubModules) + 1
	}

	subModule := models.SubModule{
		ID:          primitive.NewObjectID(),
		Name:        req.Name,
		Description: req.Description,
		Content:     req.Content,
		IsPublished: false, // Always start as draft
		Order:       order,
		CreatedAt:   now,
		UpdatedAt:   now,
		CreatedBy:   userID,
		UpdatedBy:   userID,
	}

	module.SubModules = append(module.SubModules, subModule)
	module.UpdatedAt = now
	module.UpdatedBy = userID

	if err := s.moduleRepo.UpdateModule(ctx, module); err != nil {
		return nil, fmt.Errorf("failed to create submodule: %w", err)
	}

	return &subModule, nil
}

func (s *moduleService) UpdateSubModule(ctx context.Context, moduleID primitive.ObjectID, subModuleID primitive.ObjectID, req *models.CreateSubModuleRequest, userID primitive.ObjectID) (*models.SubModule, error) {
	module, err := s.moduleRepo.GetModuleByID(ctx, moduleID)
	if err != nil {
		return nil, fmt.Errorf("module not found: %w", err)
	}

	// Find the submodule
	var subModuleIndex = -1
	for i, subModule := range module.SubModules {
		if subModule.ID == subModuleID {
			subModuleIndex = i
			break
		}
	}

	if subModuleIndex == -1 {
		return nil, fmt.Errorf("submodule not found")
	}

	// Update submodule
	module.SubModules[subModuleIndex].Name = req.Name
	module.SubModules[subModuleIndex].Description = req.Description
	module.SubModules[subModuleIndex].Content = req.Content
	if req.Order != 0 {
		module.SubModules[subModuleIndex].Order = req.Order
	}
	module.SubModules[subModuleIndex].UpdatedAt = time.Now()
	module.SubModules[subModuleIndex].UpdatedBy = userID

	module.UpdatedAt = time.Now()
	module.UpdatedBy = userID

	if err := s.moduleRepo.UpdateModule(ctx, module); err != nil {
		return nil, fmt.Errorf("failed to update submodule: %w", err)
	}

	return &module.SubModules[subModuleIndex], nil
}

func (s *moduleService) DeleteSubModule(ctx context.Context, moduleID primitive.ObjectID, subModuleID primitive.ObjectID) error {
	module, err := s.moduleRepo.GetModuleByID(ctx, moduleID)
	if err != nil {
		return fmt.Errorf("module not found: %w", err)
	}

	// Find and remove the submodule
	var newSubModules []models.SubModule
	found := false
	for _, subModule := range module.SubModules {
		if subModule.ID != subModuleID {
			newSubModules = append(newSubModules, subModule)
		} else {
			found = true
		}
	}

	if !found {
		return fmt.Errorf("submodule not found")
	}

	module.SubModules = newSubModules
	module.UpdatedAt = time.Now()

	if err := s.moduleRepo.UpdateModule(ctx, module); err != nil {
		return fmt.Errorf("failed to delete submodule: %w", err)
	}

	return nil
}

func (s *moduleService) ToggleSubModulePublication(ctx context.Context, moduleID primitive.ObjectID, subModuleID primitive.ObjectID, published bool, userID primitive.ObjectID) (*models.SubModule, error) {
	module, err := s.moduleRepo.GetModuleByID(ctx, moduleID)
	if err != nil {
		return nil, fmt.Errorf("module not found: %w", err)
	}

	// Find the submodule
	var subModuleIndex = -1
	for i, subModule := range module.SubModules {
		if subModule.ID == subModuleID {
			subModuleIndex = i
			break
		}
	}

	if subModuleIndex == -1 {
		return nil, fmt.Errorf("submodule not found")
	}

	module.SubModules[subModuleIndex].IsPublished = published
	module.SubModules[subModuleIndex].UpdatedAt = time.Now()
	module.SubModules[subModuleIndex].UpdatedBy = userID

	module.UpdatedAt = time.Now()
	module.UpdatedBy = userID

	if err := s.moduleRepo.UpdateModule(ctx, module); err != nil {
		return nil, fmt.Errorf("failed to update submodule publication status: %w", err)
	}

	return &module.SubModules[subModuleIndex], nil
}

func (s *moduleService) ReorderModules(ctx context.Context, moduleIDs []string, userID primitive.ObjectID) error {
	for i, moduleIDStr := range moduleIDs {
		moduleID, err := primitive.ObjectIDFromHex(moduleIDStr)
		if err != nil {
			return fmt.Errorf("invalid module ID %s: %w", moduleIDStr, err)
		}

		module, err := s.moduleRepo.GetModuleByID(ctx, moduleID)
		if err != nil {
			return fmt.Errorf("module not found %s: %w", moduleIDStr, err)
		}

		// Update order to match position in array (1-based)
		module.Order = i + 1
		module.UpdatedAt = time.Now()
		module.UpdatedBy = userID

		if err := s.moduleRepo.UpdateModule(ctx, module); err != nil {
			return fmt.Errorf("failed to update module order %s: %w", moduleIDStr, err)
		}
	}

	return nil
}

func (s *moduleService) ReorderSubModules(ctx context.Context, moduleID primitive.ObjectID, subModuleIDs []string, userID primitive.ObjectID) error {
	module, err := s.moduleRepo.GetModuleByID(ctx, moduleID)
	if err != nil {
		return fmt.Errorf("module not found: %w", err)
	}

	// Create a map for quick lookup of submodules
	subModuleMap := make(map[string]*models.SubModule)
	for i := range module.SubModules {
		subModuleMap[module.SubModules[i].ID.Hex()] = &module.SubModules[i]
	}

	// Reorder submodules according to the provided IDs
	var reorderedSubModules []models.SubModule
	for i, subModuleIDStr := range subModuleIDs {
		subModule, exists := subModuleMap[subModuleIDStr]
		if !exists {
			return fmt.Errorf("submodule not found: %s", subModuleIDStr)
		}

		// Update order to match position in array (1-based)
		subModule.Order = i + 1
		subModule.UpdatedAt = time.Now()
		subModule.UpdatedBy = userID

		reorderedSubModules = append(reorderedSubModules, *subModule)
	}

	// Update module with reordered submodules
	module.SubModules = reorderedSubModules
	module.UpdatedAt = time.Now()
	module.UpdatedBy = userID

	if err := s.moduleRepo.UpdateModule(ctx, module); err != nil {
		return fmt.Errorf("failed to update submodule order: %w", err)
	}

	return nil
}

func (s *moduleService) BulkReorder(ctx context.Context, req *models.BulkReorderRequest, userID primitive.ObjectID) error {
	// Handle module reordering
	if len(req.ModuleUpdates) > 0 {
		// Add user ID to each update
		for i := range req.ModuleUpdates {
			req.ModuleUpdates[i].UpdatedBy = userID
		}

		if err := s.moduleRepo.BulkUpdateModuleOrder(ctx, req.ModuleUpdates); err != nil {
			return fmt.Errorf("failed to bulk update module order: %w", err)
		}
	}

	// Handle submodule reordering
	if len(req.SubModuleUpdates) > 0 {
		// Group submodule updates by parent module
		moduleSubModules := make(map[primitive.ObjectID][]models.SubModuleOrderUpdate)

		for _, update := range req.SubModuleUpdates {
			// We need to find the parent module for each submodule
			// This requires querying each module to find which one contains the submodule
			modules, _, err := s.moduleRepo.GetAllModules(ctx, &models.GetModulesRequest{Page: 1, Limit: 1000})
			if err != nil {
				return fmt.Errorf("failed to get modules for submodule reordering: %w", err)
			}

			var parentModuleID primitive.ObjectID
			for _, module := range modules {
				for _, subModule := range module.SubModules {
					if subModule.ID == update.SubModuleID {
						parentModuleID = module.ID
						break
					}
				}
				if !parentModuleID.IsZero() {
					break
				}
			}

			if parentModuleID.IsZero() {
				return fmt.Errorf("parent module not found for submodule: %s", update.SubModuleID.Hex())
			}

			moduleSubModules[parentModuleID] = append(moduleSubModules[parentModuleID], update)
		}

		// Update each module's submodules
		for moduleID, subModuleUpdates := range moduleSubModules {
			module, err := s.moduleRepo.GetModuleByID(ctx, moduleID)
			if err != nil {
				return fmt.Errorf("failed to get module %s: %w", moduleID.Hex(), err)
			}

			// Update submodule orders
			for _, update := range subModuleUpdates {
				for i := range module.SubModules {
					if module.SubModules[i].ID == update.SubModuleID {
						module.SubModules[i].Order = update.Order
						module.SubModules[i].UpdatedAt = time.Now()
						module.SubModules[i].UpdatedBy = userID
						break
					}
				}
			}

			module.UpdatedAt = time.Now()
			module.UpdatedBy = userID

			if err := s.moduleRepo.UpdateModule(ctx, module); err != nil {
				return fmt.Errorf("failed to update module %s with reordered submodules: %w", moduleID.Hex(), err)
			}
		}
	}

	return nil
}

File: frontend/src/services/moduleService.ts
import { api } from '@/lib/api'

// Types for Module API
export interface Module {
  id: string
  name: string
  description: string
  content: string
  is_published: boolean
  order: number
  sub_modules: SubModule[]
  created_at: string
  updated_at: string
  created_by: string
  updated_by: string
}

export interface SubModule {
  id: string
  name: string
  description: string
  content: string
  is_published: boolean
  order: number
  created_at: string
  updated_at: string
  created_by: string
  updated_by: string
}

export interface CreateModuleRequest {
  name: string
  description: string
  content: string
  sub_modules?: SubModule[]
}

export interface UpdateModuleRequest {
  name?: string
  description?: string
  content?: string
  sub_modules?: SubModule[]
}

export interface CreateSubModuleRequest {
  name: string
  description: string
  content: string
}

export interface GetModulesRequest {
  page?: number
  limit?: number
  search?: string
  published?: boolean
}

export interface GetModulesResponse {
  modules: Module[]
  total: number
  page: number
  limit: number
  total_pages: number
}

// Order update interfaces for drag-and-drop functionality
export interface ModuleOrderUpdate {
  module_id: string
  order: number
}

export interface SubModuleOrderUpdate {
  submodule_id: string
  order: number
}

export interface BulkReorderRequest {
  module_updates?: ModuleOrderUpdate[]
  submodule_updates?: SubModuleOrderUpdate[]
}

// Drag and drop item interface
export interface DragDropItem {
  id: string
  type: 'module' | 'submodule'
  parentId?: string
  order: number
  title: string
  isPublished: boolean
}

// Module Service
export const moduleService = {
  // Get all modules with filtering and pagination
  getModules: async (params?: GetModulesRequest): Promise<GetModulesResponse> => {
    const searchParams = new URLSearchParams()
    
    if (params?.page) searchParams.append('page', params.page.toString())
    if (params?.limit) searchParams.append('limit', params.limit.toString())
    if (params?.search) searchParams.append('search', params.search)
    if (params?.published !== undefined) searchParams.append('published', params.published.toString())

    const query = searchParams.toString()
    return api.get<GetModulesResponse>(`/modules${query ? `?${query}` : ''}`)
  },

  // Get module by ID
  getModule: async (id: string): Promise<Module> => {
    return api.get<Module>(`/modules/${id}`)
  },

  // Create new module (Admin only)
  createModule: async (data: CreateModuleRequest): Promise<Module> => {
    return api.post<Module>('/admin/modules', data)
  },

  // Update module (Admin only)
  updateModule: async (id: string, data: UpdateModuleRequest): Promise<Module> => {
    return api.put<Module>(`/admin/modules/${id}`, data)
  },

  // Delete module (Admin only)
  deleteModule: async (id: string): Promise<void> => {
    return api.delete<void>(`/admin/modules/${id}`)
  },

  // Toggle module publication status (Admin only)
  toggleModulePublication: async (id: string, published: boolean): Promise<Module> => {
    return api.patch<Module>(`/admin/modules/${id}/publish`, { published })
  },

  // SubModule operations

  // Create submodule (Admin only)
  createSubModule: async (moduleId: string, data: CreateSubModuleRequest): Promise<SubModule> => {
    return api.post<SubModule>(`/admin/modules/${moduleId}/submodules`, data)
  },

  // Update submodule (Admin only)
  updateSubModule: async (moduleId: string, subModuleId: string, data: CreateSubModuleRequest): Promise<SubModule> => {
    return api.put<SubModule>(`/admin/modules/${moduleId}/submodules/${subModuleId}`, data)
  },

  // Delete submodule (Admin only)
  deleteSubModule: async (moduleId: string, subModuleId: string): Promise<void> => {
    return api.delete<void>(`/admin/modules/${moduleId}/submodules/${subModuleId}`)
  },

  // Toggle submodule publication status (Admin only)
  toggleSubModulePublication: async (moduleId: string, subModuleId: string, published: boolean): Promise<SubModule> => {
    return api.patch<SubModule>(`/admin/modules/${moduleId}/submodules/${subModuleId}/publish`, { published })
  },

  // Reorder modules (Admin only)
  reorderModules: async (moduleIds: string[]): Promise<void> => {
    console.log('Reordering modules:', moduleIds)
    return api.post<void>('/admin/modules/reorder', { module_ids: moduleIds })
  },

  // Reorder submodules (Admin only)
  reorderSubModules: async (moduleId: string, subModuleIds: string[]): Promise<void> => {
    console.log('Reordering submodules:', { moduleId, subModuleIds })
    return api.post<void>(`/admin/modules/${moduleId}/submodules/reorder`, { submodule_ids: subModuleIds })
  },

  // Bulk reorder modules and submodules (Admin only)
  bulkReorder: async (data: BulkReorderRequest): Promise<void> => {
    return api.post<void>('/admin/modules/bulk-reorder', data)
  },

  // Helper function to convert modules to drag-drop items
  convertToDragDropItems: (modules: Module[]): DragDropItem[] => {
    const items: DragDropItem[] = []
    
    modules.forEach(module => {
      // Add module
      items.push({
        id: module.id,
        type: 'module',
        order: module.order || 0,
        title: module.name,
        isPublished: module.is_published
      })
      
      // Add submodules
      if (module.sub_modules) {
        module.sub_modules.forEach(subModule => {
          items.push({
            id: subModule.id,
            type: 'submodule',
            parentId: module.id,
            order: subModule.order || 0,
            title: subModule.name,
            isPublished: subModule.is_published
          })
        })
      }
    })
    
    return items.sort((a, b) => {
      // Sort modules first, then submodules within each module
      if (a.type === 'module' && b.type === 'module') {
        return a.order - b.order
      }
      if (a.type === 'submodule' && b.type === 'submodule') {
        if (a.parentId === b.parentId) {
          return a.order - b.order
        }
        // Different parents, maintain module order
        const moduleA = modules.find(m => m.id === a.parentId)
        const moduleB = modules.find(m => m.id === b.parentId)
        return (moduleA?.order || 0) - (moduleB?.order || 0)
      }
      // Module comes before its submodules
      if (a.type === 'module' && b.type === 'submodule' && b.parentId === a.id) {
        return -1
      }
      if (b.type === 'module' && a.type === 'submodule' && a.parentId === b.id) {
        return 1
      }
      // General module vs submodule ordering
      return a.type === 'module' ? -1 : 1
    })
  },
}

// Hook for easier use in components (optional - using React Query would be better)
export const useModules = () => {
  return {
    modules: [],
    loading: false,
    error: null,
    refetch: () => Promise.resolve()
  }
} 